concordium-cis2/src/cis2_client.rs

```diff
@@ -0,0 +1,696 @@
+    /// let mut host = TestHost::new((), TestStateBuilder::new());
+    /// host.setup_mock_entrypoint(
+    ///     ContractAddress::new(0, 0),
+    ///     OwnedEntrypointName::new_unchecked("updateOperator".to_string()),
+    ///     MockFn::new_v1(|_, _, _, _| Ok((false, ()))),
+    /// );
+    ///
+    /// let client = Cis2Client::new(ContractAddress::new(0, 0));
+    /// let res: Result<bool, Cis2ClientError<()>> = client.update_operator(
+    ///     &mut host,
+    ///     Address::Account(AccountAddress([1; 32])),
+    ///     OperatorUpdate::Add,
+    /// );
+    ///
+    /// assert!(res.is_ok());
+    /// ```
+    pub fn update_operator<State, E: Deserial>(
+        &self,
+        host: &mut impl HasHost<State>,
+        operator: Address,
+        update: OperatorUpdate,
+    ) -> Result<bool, Cis2ClientError<E>> {
+        let params = UpdateOperator {
+            operator,
+            update,
+        };
+        let (state_modified, _): (bool, Option<()>) =
+            self.invoke_contract(host, UPDATE_OPERATOR_ENTRYPOINT_NAME, &params)?;
+
+        Ok(state_modified)
+    }
+
+    fn invoke_contract_read_only<State, P: Serial, R: Deserial, E: Deserial>(
+        &self,
+        host: &impl HasHost<State>,
+        method: EntrypointName,
+        parameter: &P,
+    ) -> Result<R, Cis2ClientError<E>> {
+        let res =
+            host.invoke_contract_read_only(&self.contract, parameter, method, Amount::from_ccd(0));
+
+        let res = match res {
+            Ok(val) => val,
+            Err(err) => return Err(Cis2ClientError::<E>::try_from(err)?),
+        };
+
+        let res = match res {
+            // Since the contract should return a response. If it doesn't, it is an error.
+            Some(mut res) => R::deserial(&mut res)?,
+            None => bail!(Cis2ClientError::InvalidResponse),
+        };
+
+        Ok(res)
+    }
+
+    fn invoke_contract<State, P: Serial, R: Deserial, E: Deserial>(
+        &self,
+        host: &mut impl HasHost<State>,
+        method: EntrypointName,
+        parameter: &P,
+    ) -> Result<(bool, Option<R>), Cis2ClientError<E>> {
+        let res = host.invoke_contract(&self.contract, parameter, method, Amount::from_ccd(0));
+
+        let res = match res {
+            Ok(val) => {
+                let o = match val.1 {
+                    Some(mut res) => Some(R::deserial(&mut res)?),
+                    None => None,
+                };
+                (val.0, o)
+            }
+            Err(err) => return Err(Cis2ClientError::<E>::try_from(err)?),
+        };
+
+        Ok(res)
+    }
+
+    fn first<T, E>(array: &mut Vec<T>) -> Result<T, Cis2ClientError<E>> {
+        // If the contract returns a response, but the response is empty, it is an
+        // error. Since for a single query the response should be non-empty.
+        ensure!(!array.is_empty(), Cis2ClientError::InvalidResponse);
+
+        Ok(array.swap_remove(0))
+    }
```

Comment:
This should not be a method, it does not construct `Self` nor does it use it.

Reply:
Also, thsi seems to only be used for single queries. In that case it should ensure there is exactly one value in the response.

Reply:
removed